airflow/providers/amazon/aws/secrets/ssm.py

```diff
@@ -0,0 +1,120 @@
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+"""
+Objects relating to sourcing connections from AWS SSM Parameter Store
+"""
+
+from typing import List, Optional
+
+import boto3
+
+from airflow.configuration import conf
+from airflow.models import Connection
+from airflow.secrets import CONN_ENV_PREFIX, BaseSecretsBackend
+
+
+class AwsSsmSecretsBackend(BaseSecretsBackend):
+    """
+    Retrieves Connection object from AWS SSM Parameter Store
+
+    Configurable via airflow config like so:
+
+    .. code-block:: ini
+
+        [secrets_backend]
+        class_list = airflow.providers.amazon.aws.secrets.ssm.AwsSsmSecretsBackend
+        aws_ssm_prefix = /airflow
+        aws_ssm_profile_name = default
+
+    For example, if ssm path is ``/airflow/AIRFLOW_CONN_SMTP_DEFAULT``, this would be accessible if you
+    provide ``aws_ssm_prefix = /airflow`` and conn_id ``smtp_default``.
+
+    """
+
+    CONF_SECTION = "secrets_backend"
+    CONF_KEY_SSM_PREFIX = "aws_ssm_prefix"
+    DEFAULT_PREFIX = "/airflow"
+    CONF_KEY_PROFILE_NAME = "aws_ssm_profile_name"
+
+    def __init__(self, *args, **kwargs):
+        pass
+
+    @property
+    def ssm_prefix(self) -> str:
+        """
+        Gets ssm prefix from conf.
+
+        Ensures that there is no trailing slash.
+
+        :return:
+        """
+        ssm_prefix = conf.get(
+            section=self.CONF_SECTION, key=self.CONF_KEY_SSM_PREFIX, fallback=self.DEFAULT_PREFIX
+        )
+        return ssm_prefix.rstrip("/")
+
+    @property
+    def aws_profile_name(self) -> Optional[str]:
+        """
+        Gets AWS profile to use from conf.
+
+        :return:
+        """
+        profile_name = conf.get(
+            section=self.CONF_SECTION, key=self.CONF_KEY_PROFILE_NAME, fallback=None
+        )
+        return profile_name or None
+
+    def build_ssm_path(self, conn_id: str):
+        """
+        Given conn_id, build SSM path.
+
+        Assumes connection params use same naming convention as env vars, but may have arbitrary prefix.
+
+        :param conn_id:
+        :return:
+        """
+        param_name = (CONN_ENV_PREFIX + conn_id).upper()
+        param_path = self.ssm_prefix + "/" + param_name
+        return param_path
+
+    def get_conn_uri(self, conn_id):
+        """
+        Get param value
+
+        :param conn_id:
+        :return:
+        """
+        session = boto3.Session(profile_name=self.aws_profile_name)
+        client = session.client("ssm")
+        response = client.get_parameter(
+            Name=self.build_ssm_path(conn_id=conn_id), WithDecryption=True
+        )
+        value = response["Parameter"]["Value"]
+        return value
+
+    def get_connections(self, conn_id) -> List[Connection]:
```

Comment:
@kaxil @potiuk should all methods be class methods (i.e. both base and implementations)?  There isn't really any need to have any instances since they can't be instantiated in a dynamic way, right?

Reply:
Yes, please, if we don't need instances, just make it class method

Reply:
@kaxil looked into it... if i make everything class methods, i can't use `property` decorator....

e.g. here: https://github.com/apache/airflow/pull/6376/files#diff-2e589119610e1b6ffe1ff76755881a4bR58

what would you recommend i do?  there is no built-in class property in python, but i have seen implementations. options:
* forget about switching to class methods
* convert properties to `get_some_thing()` class methods
* use custom classproperty decorator